File: v4/rust/iprange-livedb/src/worker/wire_cleanup.rs
use std::path::{Path, PathBuf};

use crate::error::{Error, Result};
use crate::publication::cleanup::EarlyDiscard;
use crate::publication::{CleanupArtifacts, CreationSecurity, PublicationProblem};
use crate::recovery::{ScratchCleanup, ScratchProblem, ScratchResidue};

use super::control::{Control, ScratchCheckpoint, ScratchCheckpointEntry};
use super::wire::{self, Reader, Writer};
use super::wire_publication;

pub(super) struct Request {
    pub(super) destination_path: PathBuf,
    pub(super) output: crate::publication::PrivateOutputAttempt,
    pub(super) scratch_directory: Option<PathBuf>,
    pub(super) scratch: Option<ScratchCheckpoint>,
}

pub(super) fn write_request(
    control: &Control,
    destination_path: &Path,
    output_attempt: &crate::publication::PrivateOutputAttempt,
    scratch_directory: Option<&Path>,
    scratch: Option<&ScratchCheckpoint>,
) -> Result<()> {
    let mut output = Writer::new(control);
    output.path(destination_path)?;
    wire_publication::private_output(&mut output, output_attempt)?;
    output.optional_path(scratch_directory)?;
    write_checkpoint(&mut output, scratch)?;
    output.finish()
}

pub(super) fn read_request(control: &Control) -> Result<Request> {
    let mut input = Reader::new(control)?;
    let request = Request {
        destination_path: input.path()?,
        output: wire_publication::read_private_output(&mut input)?,
        scratch_directory: input.optional_path()?,
        scratch: read_checkpoint(&mut input)?,
    };
    input.finish()?;
    if request.scratch.is_some() != request.scratch_directory.is_some() {
        return Err(Error::Corrupt(
            "worker cleanup scratch path and checkpoint disagree",
        ));
    }
    Ok(request)
}

pub(super) fn write_result(
    control: &Control,
    discarded: &EarlyDiscard,
    scratch: Option<&ScratchCleanup>,
) -> Result<()> {
    let mut output = Writer::new(control);
    wire_publication::private_output(&mut output, &discarded.output)?;
    let mut artifacts = CleanupArtifacts::new();
    if let Some(artifact) = &discarded.artifact {
        artifacts.push(artifact.clone());
    }
    wire_publication::cleanup(&mut output, &artifacts)?;
    output.byte(wire_publication::housekeeping_value(discarded.housekeeping))?;
    wire_publication::write_housekeeping_list(&mut output, &discarded.visible_housekeeping)?;
    write_scratch(&mut output, scratch)?;
    output.finish()
}

pub(super) fn read_result(control: &Control) -> Result<(EarlyDiscard, Option<ScratchCleanup>)> {
    let mut input = Reader::new(control)?;
    let output = wire_publication::read_private_output(&mut input)?;
    let artifacts = wire_publication::read_cleanup(&mut input)?;
    if artifacts.len() > 1 {
        return Err(Error::Corrupt(
            "worker cleanup returned multiple output residues",
        ));
    }
    let discarded = EarlyDiscard {
        output,
        artifact: artifacts.get(0).cloned(),
        housekeeping: wire_publication::read_housekeeping_value(input.byte()?)?,
        visible_housekeeping: wire_publication::read_housekeeping_artifacts(&mut input)?,
    };
    let scratch = read_scratch(&mut input)?;
    input.finish()?;
    Ok((discarded, scratch))
}

fn write_checkpoint(output: &mut Writer<'_>, checkpoint: Option<&ScratchCheckpoint>) -> Result<()> {
    output.bool(checkpoint.is_some())?;
    let Some(checkpoint) = checkpoint else {
        return Ok(());
    };
    output.bytes(&checkpoint.attempt_id)?;
    wire::identity(output, checkpoint.directory_identity)?;
    output.u16(checkpoint.creation_security.kind)?;
    output.bytes(&checkpoint.creation_security.commitment)?;
    output.byte(
        u8::try_from(checkpoint.entries.len())
            .map_err(|_| Error::BudgetExceeded("worker scratch checkpoint entries"))?,
    )?;
    for entry in &checkpoint.entries {
        output.u32(entry.ordinal)?;
        wire::identity(output, entry.identity)?;
    }
    Ok(())
}

fn read_checkpoint(input: &mut Reader<'_>) -> Result<Option<ScratchCheckpoint>> {
    if !input.bool()? {
        return Ok(None);
    }
    let attempt_id = input.array()?;
    let directory_identity = wire::read_identity(input)?;
    let creation_security = CreationSecurity {
        kind: input.u16()?,
        commitment: input.array()?,
    };
    let count = input.byte()? as usize;
    if attempt_id == [0; 16]
        || !valid_identity(directory_identity)
        || !valid_security(creation_security.kind, creation_security.commitment)
        || count > 2
    {
        return Err(Error::Corrupt("worker scratch checkpoint is invalid"));
    }
    let mut entries = Vec::with_capacity(count);
    for _ in 0..count {
        let entry = ScratchCheckpointEntry {
            ordinal: input.u32()?,
            identity: wire::read_identity(input)?,
        };
        if !valid_identity(entry.identity)
            || entries.iter().any(|prior: &ScratchCheckpointEntry| {
                prior.ordinal == entry.ordinal || prior.identity == entry.identity
            })
        {
            return Err(Error::Corrupt(
                "worker scratch checkpoint contains duplicate authority",
            ));
        }
        entries.push(entry);
    }
    Ok(Some(ScratchCheckpoint {
        attempt_id,
        directory_identity,
        creation_security,
        entries,
    }))
}

fn write_scratch(output: &mut Writer<'_>, cleanup: Option<&ScratchCleanup>) -> Result<()> {
    output.bool(cleanup.is_some())?;
    let Some(cleanup) = cleanup else {
        return Ok(());
    };
    output.bytes(&cleanup.attempt_id)?;
    wire::identity(output, cleanup.directory_identity)?;
    output.u16(cleanup.creation_security_kind)?;
    output.bytes(&cleanup.creation_security_commitment)?;
    output.byte(
        u8::try_from(cleanup.residues.len())
            .map_err(|_| Error::BudgetExceeded("worker scratch cleanup residues"))?,
    )?;
    for residue in &cleanup.residues {
        output.u32(residue.ordinal)?;
        wire::identity(output, residue.directory_identity)?;
        output.sized_bytes(&residue.basename)?;
        wire::identity(output, residue.identity)?;
        output.u16(residue.creation_security_kind)?;
        output.bytes(&residue.creation_security_commitment)?;
        wire_publication::problem(
            output,
            &PublicationProblem {
                code: residue.problem.code,
                os_code: residue.problem.os_code,
                detail: residue.problem.detail.clone(),
            },
        )?;
    }
    output.byte(wire_publication::housekeeping_value(cleanup.housekeeping))?;
    wire_publication::write_housekeeping_list(output, &cleanup.visible_housekeeping)
}

fn read_scratch(input: &mut Reader<'_>) -> Result<Option<ScratchCleanup>> {
    if !input.bool()? {
        return Ok(None);
    }
    let attempt_id = input.array()?;
    let directory_identity = wire::read_identity(input)?;
    let creation_security_kind = input.u16()?;
    let creation_security_commitment = input.array()?;
    let count = input.byte()? as usize;
    if attempt_id == [0; 16]
        || !valid_identity(directory_identity)
        || !valid_security(creation_security_kind, creation_security_commitment)
        || count > 2
    {
        return Err(Error::Corrupt("worker scratch cleanup is invalid"));
    }
    let mut residues = Vec::with_capacity(count);
    for _ in 0..count {
        let ordinal = input.u32()?;
        let residue_directory = wire::read_identity(input)?;
        let basename = input.boxed_bytes()?;
        let identity = wire::read_identity(input)?;
        let security_kind = input.u16()?;
        let security_commitment = input.array()?;
        let problem = wire_publication::read_problem(input)?;
        let expected_basename = crate::recovery::checkpoint_basename(attempt_id, ordinal)?;
        if residue_directory != directory_identity
            || !valid_identity(identity)
            || security_kind != creation_security_kind
            || security_commitment != creation_security_commitment
            || basename != expected_basename
            || residues.iter().any(|prior: &ScratchResidue| {
                prior.ordinal == ordinal || prior.identity == identity
            })
        {
            return Err(Error::Corrupt(
                "worker scratch cleanup authority is inconsistent",
            ));
        }
        residues.push(ScratchResidue {
            ordinal,
            directory_identity: residue_directory,
            basename,
            identity,
            creation_security_kind: security_kind,
            creation_security_commitment: security_commitment,
            problem: ScratchProblem {
                code: problem.code,
                os_code: problem.os_code,
                detail: problem.detail,
            },
        });
    }
    Ok(Some(ScratchCleanup {
        attempt_id,
        directory_identity,
        creation_security_kind,
        creation_security_commitment,
        residues,
        housekeeping: wire_publication::read_housekeeping_value(input.byte()?)?,
        visible_housekeeping: wire_publication::read_housekeeping_artifacts(input)?.into_vec(),
    }))
}

fn valid_identity(identity: crate::validation::LocalFileIdentity) -> bool {
    identity.kind == crate::publication::namespace::IDENTITY_KIND
        && crate::publication::namespace::Identity::decode(identity.bytes).is_some()
}

fn valid_security(kind: u16, commitment: [u8; 32]) -> bool {
    kind == crate::publication::namespace::CREATION_SECURITY_KIND && commitment != [0; 32]
}

File: v4/go/internal/worker/wire_cleanup.go
//go:build linux && amd64

// Cleanup-mode wire codecs (Rust worker/wire_cleanup.rs): the
// cleanup-guard request (destination path, private output attempt,
// optional scratch directory and checkpoint) and the discard result
// (the early-discard facts, the optional scratch cleanup). Every field
// order and every consistency cross-check mirrors the Rust authority;
// the checkpoint and scratch encodings repeat the control-page
// validation exactly as the Rust wire layer does.

package worker

import (
	"bytes"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/publication"
)

// CleanupRequest is the decoded cleanup request (Rust
// wire_cleanup.rs Request).
type CleanupRequest struct {
	DestinationPath  string
	Output           publication.PrivateOutputAttempt
	ScratchDirectory *string
	Scratch          *ScratchCheckpoint
}

// WriteCleanupRequest writes one cleanup request (Rust
// wire_cleanup::write_request: destination path, the private output
// attempt, the optional scratch directory, and the optional scratch
// checkpoint).
func WriteCleanupRequest(control *Control, destinationPath string, output *publication.PrivateOutputAttempt, scratchDirectory *string, scratch *ScratchCheckpoint) error {
	w := NewWireWriter(control)
	if err := w.Path(destinationPath); err != nil {
		return err
	}
	if err := writePrivateOutput(w, output); err != nil {
		return err
	}
	if err := w.OptionalPath(scratchDirectory); err != nil {
		return err
	}
	if err := writeScratchCheckpoint(w, scratch); err != nil {
		return err
	}
	return w.Finish()
}

// ReadCleanupRequest decodes one cleanup request (Rust
// wire_cleanup::read_request: the scratch path and checkpoint must
// agree).
func ReadCleanupRequest(control *Control) (*CleanupRequest, error) {
	r, err := NewWireReader(control)
	if err != nil {
		return nil, err
	}
	request := &CleanupRequest{}
	if request.DestinationPath, err = r.Path(); err != nil {
		return nil, err
	}
	if request.Output, err = readPrivateOutput(r); err != nil {
		return nil, err
	}
	if request.ScratchDirectory, err = r.OptionalPath(); err != nil {
		return nil, err
	}
	if request.Scratch, err = readScratchCheckpoint(r); err != nil {
		return nil, err
	}
	if err := r.Finish(); err != nil {
		return nil, err
	}
	if (request.Scratch != nil) != (request.ScratchDirectory != nil) {
		return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker cleanup scratch path and checkpoint disagree"}
	}
	return request, nil
}

// EarlyDiscard is the wire form of the pre-publication discard facts
// (Rust publication::cleanup::EarlyDiscard): the discarded private
// output attempt, the optional unresolved artifact, and the
// housekeeping evidence. The Go publication machine keeps its discard
// facts private, so the worker boundary owns the wire shape.
type EarlyDiscard struct {
	Output              publication.PrivateOutputAttempt
	Artifact            *publication.CleanupArtifact
	Housekeeping        publication.Housekeeping
	VisibleHousekeeping []publication.HousekeepingArtifact
}

// Clean reports whether the discard left no residue at all (Rust
// client discard_clean).
func (d *EarlyDiscard) Clean() bool {
	return d.Artifact == nil && d.Housekeeping == publication.HousekeepingNone && len(d.VisibleHousekeeping) == 0
}

// WriteCleanupResult writes one cleanup result (Rust
// wire_cleanup::write_result: the discarded output, its optional
// artifact as a cleanup ledger, the housekeeping facts, and the
// optional scratch cleanup).
func WriteCleanupResult(control *Control, discarded *EarlyDiscard, scratch *ScratchCleanup) error {
	w := NewWireWriter(control)
	if err := writePrivateOutput(w, &discarded.Output); err != nil {
		return err
	}
	artifacts := publication.NewCleanupArtifacts()
	if discarded.Artifact != nil {
		artifacts.Push(*discarded.Artifact)
	}
	if err := writeCleanupArtifacts(w, &artifacts); err != nil {
		return err
	}
	if err := writeHousekeepingValue(w, discarded.Housekeeping); err != nil {
		return err
	}
	if err := writeHousekeepingList(w, discarded.VisibleHousekeeping); err != nil {
		return err
	}
	if err := writeScratchCleanup(w, scratch); err != nil {
		return err
	}
	return w.Finish()
}

// ReadCleanupResult decodes one cleanup result (Rust
// wire_cleanup::read_result: more than one output residue is
// corruption).
func ReadCleanupResult(control *Control) (*EarlyDiscard, *ScratchCleanup, error) {
	r, err := NewWireReader(control)
	if err != nil {
		return nil, nil, err
	}
	discarded := &EarlyDiscard{}
	if discarded.Output, err = readPrivateOutput(r); err != nil {
		return nil, nil, err
	}
	artifacts, err := readCleanupArtifacts(r)
	if err != nil {
		return nil, nil, err
	}
	if artifacts.Len() > 1 {
		return nil, nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker cleanup returned multiple output residues"}
	}
	if artifacts.Len() == 1 {
		discarded.Artifact = artifacts.At(0)
	}
	housekeeping, err := r.Byte()
	if err != nil {
		return nil, nil, err
	}
	if discarded.Housekeeping, err = readHousekeepingValueByte(housekeeping); err != nil {
		return nil, nil, err
	}
	if discarded.VisibleHousekeeping, err = readHousekeepingArtifacts(r); err != nil {
		return nil, nil, err
	}
	scratch, err := readScratchCleanup(r)
	if err != nil {
		return nil, nil, err
	}
	if err := r.Finish(); err != nil {
		return nil, nil, err
	}
	return discarded, scratch, nil
}

// writeScratchCheckpoint writes one optional scratch checkpoint (Rust
// wire_cleanup::write_checkpoint: presence bit, then the attempt
// identity, the directory and security facts, and the bounded entry
// list).
func writeScratchCheckpoint(w *WireWriter, checkpoint *ScratchCheckpoint) error {
	if checkpoint == nil {
		return w.Bool(false)
	}
	if err := w.Bool(true); err != nil {
		return err
	}
	if err := w.Bytes(checkpoint.AttemptID[:]); err != nil {
		return err
	}
	if err := writeIdentity(w, checkpoint.DirectoryIdentity); err != nil {
		return err
	}
	if err := w.U16(checkpoint.CreationSecurity.Kind); err != nil {
		return err
	}
	if err := w.Bytes(checkpoint.CreationSecurity.Commitment[:]); err != nil {
		return err
	}
	if len(checkpoint.Entries) > scratchEntryCapacity {
		return &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "worker scratch checkpoint entries"}
	}
	if err := w.Byte(uint8(len(checkpoint.Entries))); err != nil {
		return err
	}
	for _, entry := range checkpoint.Entries {
		if err := w.U32(entry.Ordinal); err != nil {
			return err
		}
		if err := writeIdentity(w, entry.Identity); err != nil {
			return err
		}
	}
	return nil
}

// readScratchCheckpoint decodes one optional scratch checkpoint (Rust
// wire_cleanup::read_checkpoint: zero attempt, invalid identity or
// security, or an entry count above the capacity is corruption).
func readScratchCheckpoint(r *WireReader) (*ScratchCheckpoint, error) {
	present, err := r.Bool()
	if err != nil {
		return nil, err
	}
	if !present {
		return nil, nil
	}
	checkpoint := &ScratchCheckpoint{}
	if checkpoint.AttemptID, err = r.Array16(); err != nil {
		return nil, err
	}
	if checkpoint.DirectoryIdentity, err = readIdentity(r); err != nil {
		return nil, err
	}
	if checkpoint.CreationSecurity.Kind, err = r.U16(); err != nil {
		return nil, err
	}
	if checkpoint.CreationSecurity.Commitment, err = r.Array32(); err != nil {
		return nil, err
	}
	count, err := r.Byte()
	if err != nil {
		return nil, err
	}
	if checkpoint.AttemptID == [16]byte{} ||
		!scratchIdentityValid(checkpoint.DirectoryIdentity) ||
		!scratchSecurityValid(checkpoint.CreationSecurity.Kind, checkpoint.CreationSecurity.Commitment) ||
		count > scratchEntryCapacity {
		return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker scratch checkpoint is invalid"}
	}
	for index := 0; index < int(count); index++ {
		entry := ScratchCheckpointEntry{}
		if entry.Ordinal, err = r.U32(); err != nil {
			return nil, err
		}
		if entry.Identity, err = readIdentity(r); err != nil {
			return nil, err
		}
		if !scratchIdentityValid(entry.Identity) {
			return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker scratch checkpoint is invalid"}
		}
		for _, prior := range checkpoint.Entries {
			if prior.Ordinal == entry.Ordinal || prior.Identity == entry.Identity {
				return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker scratch checkpoint contains duplicate authority"}
			}
		}
		checkpoint.Entries = append(checkpoint.Entries, entry)
	}
	return checkpoint, nil
}

// ScratchProblem is the fixed cleanup problem of one scratch residue
// (Rust recovery::scratch::ScratchProblem: the code class, the
// optional errno, and the exact detail).
type ScratchProblem struct {
	Code   format.ErrorCode
	OSCode *int32
	Detail string
}

// ScratchResidue is one authorized scratch artifact whose durable
// absence was not proved (Rust recovery::scratch::ScratchResidue).
type ScratchResidue struct {
	Ordinal                    uint32
	DirectoryIdentity          publication.LocalFileIdentity
	Basename                   []byte
	Identity                   publication.LocalFileIdentity
	CreationSecurityKind       uint16
	CreationSecurityCommitment [32]byte
	Problem                    ScratchProblem
}

// ScratchCleanup is the terminal facts of one scratch attempt cleanup
// (Rust recovery::scratch::ScratchCleanup).
type ScratchCleanup struct {
	AttemptID                  [16]byte
	DirectoryIdentity          publication.LocalFileIdentity
	CreationSecurityKind       uint16
	CreationSecurityCommitment [32]byte
	Residues                   []ScratchResidue
	Housekeeping               publication.Housekeeping
	VisibleHousekeeping        []publication.HousekeepingArtifact
}

// Clean reports whether the scratch cleanup proved every artifact
// absent (Rust ScratchCleanup::clean plus the client scratch_clean
// housekeeping check).
func (s *ScratchCleanup) Clean() bool {
	return len(s.Residues) == 0 &&
		s.Housekeeping == publication.HousekeepingNone &&
		len(s.VisibleHousekeeping) == 0
}

// writeScratchCleanup writes one optional scratch cleanup (Rust
// wire_cleanup::write_scratch: the attempt identity, the directory and
// security facts, the bounded residue list with the checkpoint-name
// proof, the housekeeping facts).
func writeScratchCleanup(w *WireWriter, cleanup *ScratchCleanup) error {
	if cleanup == nil {
		return w.Bool(false)
	}
	if err := w.Bool(true); err != nil {
		return err
	}
	if err := w.Bytes(cleanup.AttemptID[:]); err != nil {
		return err
	}
	if err := writeIdentity(w, cleanup.DirectoryIdentity); err != nil {
		return err
	}
	if err := w.U16(cleanup.CreationSecurityKind); err != nil {
		return err
	}
	if err := w.Bytes(cleanup.CreationSecurityCommitment[:]); err != nil {
		return err
	}
	if len(cleanup.Residues) > scratchEntryCapacity {
		return &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "worker scratch cleanup residues"}
	}
	if err := w.Byte(uint8(len(cleanup.Residues))); err != nil {
		return err
	}
	for _, residue := range cleanup.Residues {
		if err := w.U32(residue.Ordinal); err != nil {
			return err
		}
		if err := writeIdentity(w, residue.DirectoryIdentity); err != nil {
			return err
		}
		if err := w.SizedBytes(residue.Basename); err != nil {
			return err
		}
		if err := writeIdentity(w, residue.Identity); err != nil {
			return err
		}
		if err := w.U16(residue.CreationSecurityKind); err != nil {
			return err
		}
		if err := w.Bytes(residue.CreationSecurityCommitment[:]); err != nil {
			return err
		}
		problem := WireProblem{Code: residue.Problem.Code, OSCode: residue.Problem.OSCode, Detail: residue.Problem.Detail}
		if err := writeProblem(w, &problem); err != nil {
			return err
		}
	}
	if err := writeHousekeepingValue(w, cleanup.Housekeeping); err != nil {
		return err
	}
	return writeHousekeepingList(w, cleanup.VisibleHousekeeping)
}

// readScratchCleanup decodes one optional scratch cleanup (Rust
// wire_cleanup::read_scratch: every residue must carry the recorded
// directory and security facts, the exact checkpoint basename, and no
// duplicate authority).
func readScratchCleanup(r *WireReader) (*ScratchCleanup, error) {
	present, err := r.Bool()
	if err != nil {
		return nil, err
	}
	if !present {
		return nil, nil
	}
	cleanup := &ScratchCleanup{}
	if cleanup.AttemptID, err = r.Array16(); err != nil {
		return nil, err
	}
	if cleanup.DirectoryIdentity, err = readIdentity(r); err != nil {
		return nil, err
	}
	if cleanup.CreationSecurityKind, err = r.U16(); err != nil {
		return nil, err
	}
	if cleanup.CreationSecurityCommitment, err = r.Array32(); err != nil {
		return nil, err
	}
	count, err := r.Byte()
	if err != nil {
		return nil, err
	}
	if cleanup.AttemptID == [16]byte{} ||
		!scratchIdentityValid(cleanup.DirectoryIdentity) ||
		!scratchSecurityValid(cleanup.CreationSecurityKind, cleanup.CreationSecurityCommitment) ||
		count > scratchEntryCapacity {
		return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker scratch cleanup is invalid"}
	}
	for index := 0; index < int(count); index++ {
		residue := ScratchResidue{}
		if residue.Ordinal, err = r.U32(); err != nil {
			return nil, err
		}
		if residue.DirectoryIdentity, err = readIdentity(r); err != nil {
			return nil, err
		}
		if residue.Basename, err = r.BoxedBytes(); err != nil {
			return nil, err
		}
		if residue.Identity, err = readIdentity(r); err != nil {
			return nil, err
		}
		if residue.CreationSecurityKind, err = r.U16(); err != nil {
			return nil, err
		}
		if residue.CreationSecurityCommitment, err = r.Array32(); err != nil {
			return nil, err
		}
		problem, err := readProblem(r)
		if err != nil {
			return nil, err
		}
		residue.Problem = ScratchProblem{Code: problem.Code, OSCode: problem.OSCode, Detail: problem.Detail}
		expected := checkpointBasename(cleanup.AttemptID, residue.Ordinal)
		if residue.DirectoryIdentity != cleanup.DirectoryIdentity ||
			!scratchIdentityValid(residue.Identity) ||
			residue.CreationSecurityKind != cleanup.CreationSecurityKind ||
			residue.CreationSecurityCommitment != cleanup.CreationSecurityCommitment ||
			!bytes.Equal(residue.Basename, expected) {
			return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker scratch cleanup authority is inconsistent"}
		}
		for _, prior := range cleanup.Residues {
			if prior.Ordinal == residue.Ordinal || prior.Identity == residue.Identity {
				return nil, &format.Error{Code: format.CodeFormatInvalid, Detail: "worker scratch cleanup authority is inconsistent"}
			}
		}
		cleanup.Residues = append(cleanup.Residues, residue)
	}
	housekeeping, err := r.Byte()
	if err != nil {
		return nil, err
	}
	if cleanup.Housekeeping, err = readHousekeepingValueByte(housekeeping); err != nil {
		return nil, err
	}
	if cleanup.VisibleHousekeeping, err = readHousekeepingArtifacts(r); err != nil {
		return nil, err
	}
	return cleanup, nil
}
